### PR TITLE
unify overload groups in the graph collector

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
Bug/issue #, if applicable: rdar://126886432

## Summary

When symbol graphs with generated overload group symbols are collected into a unified symbol graph, the resulting graph can have inconsistent information depending on how the symbols are represented across the different symbol graph. In situations where some symbols are absent or present in only some of the constituent symbol graphs, there can be different sets of overload relations or different views of the overload indices depending on which graph selector you use.

To resolve this, this PR does the following:

1. Introduces a "unified mixin" property to the unified symbol type.
2. Adds a step to `GraphCollector.finishLoading()` that recalculates overload groups if any are present:
   1. Collect all overload groups that refer to the same set of symbols.
   2. Resorts the symbols based on the ordering that the original overload group method uses.
   3. Keeps the overload group based on the first symbol in the resulting ordering, and ensures that all the symbols have relations that point to this selected group.
   4. Store the newly computed indices and overload group in the overloaded symbols' `unifiedMixins` property.
   5. If there were multiple overload groups that referred to these symbols, remove the extra "sibling" overload groups and any relations they had.

The resulting view of the symbol graph creates a consistent overload group that references all the symbols across all the input symbol graphs, via the ensuring references and unified overload data.

## Dependencies

None, though a separate PR to Swift-DocC (forthcoming) is required to properly integrate this change.

## Testing

Full testing instructions will be available in the Swift-DocC PR.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [ n/a ] Updated documentation if necessary

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
@swift-ci Please test

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (patshaughnessy) [inline Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:281]: This is a bit confusing to read  - maybe set `allSymbolsHaveDeclaration` directly using `allSatisfy` and break things up a bit... something like:

```suggestion
            let overloads = overloadedSymbols.map{ symbols[$0]! }
            let allSymbolsHaveDeclaration = overloads.allSatisfy {
                $0.mixins.values.contains(
                    where: { $0.keys.contains(SymbolGraph.Symbol.DeclarationFragments.mixinKey) }
                )
            }
            let sortedOverloads = overloads.sorted(by: Symbol.sortForOverloads(orderByDeclaration: allSymbolsHaveDeclaration))
```

... although writing this makes me think we might want to rename `overloadedSymbols` to be `overloadedSymbolIdentifiers`. It's a bit confusing to have to distinguish between symbols, overloadedSymbols etc. when one is a symbol and the other a string identifier.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (patshaughnessy) [inline Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:310]: Any performance concerns with this lookup? Should we be using a set instead of an array for `relationshipsByLanguage`'s values?

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (patshaughnessy) [inline Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:238]: It might be worth adding an assertion here to check that each target, i.e. each overloadGroup, uses `SymbolGraph.Symbol.overloadGroupIdentifierSuffix`. This would also give the reader a clue of why the assert below should always be true: `siblingOverloadGroups.contains(computedOverloadGroup)`

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (patshaughnessy) [inline Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:339]: Same here - if we have 10,000s of symbols will this be very slow?

**Review by patshaughnessy — COMMENTED** [review]
Great work! Love the detailed explanation. I just have a few minor suggestions about making the code a bit easier to read and understand.

One question before I approve this: How can it be valid to discard the sibling overload groups? (Step 5)

Maybe add a bit more explanation around this detail for future readers; thanks!

File: Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift:165]: The reason i went for this version was so that i could have one single `sorted(by:)` call where it's needed and perform the branching here in the underlying implementation. If there are separate `sortForOverloadsByDeclaration` and `sortForOverloadsByIdentifier` calls, the callsites still look like this, where a variable is initialized within an if-branch rather than inline:

https://github.com/apple/swift-docc-symbolkit/blob/8a740b5197ece51e4b597f8a467546da66324581/Sources/SymbolKit/SymbolGraph/SymbolGraph%2BOverloads.swift#L54-L59

I can see the utility of not force-unwrapping here, but it acts like a precondition: If we are trying to sort by declaration, there _must_ be a declaration available to sort by. I don't want to fall back to sorting by identifier at this point in the stack because it can create inconsistent or confusing results in the resulting sort. I can create an actual `precondition(_:_:)` call so that it emits a real error message in case of misuse, but i believe that removing the force-unwrap entirely will cause more problems than it solves.

> I wonder also if there's a generally simpler way to express this logic, not having to pass in the two different closures with the boolean flag which is a bit awkward.

I can see about simplifying it with a protocol, but by having the actual comparison in a single place like this, i can guarantee that the sorting will behave the same between single symbols and unified symbols, which was the cause of some bugs as i was testing this code.

File: Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift:194]: > Isn't `internal` the default? Can we drop that keyword here?

It is the default, but i like being explicit when the keyword is available. 😅 I think @d-ronnqvist also called out the use of `internal` in a previous PR, so i can drop them here.

> Also, I haven't seen this style before: Is this equivalent to [...]

Yes! I think this syntax was added in one of the Swift 5 point releases, from around the time i was starting to use it.

> Which is preferred from a Swift style point of view?

That one i don't know. I very much like this sugar since you almost never write the `Array` type name directly and this makes it much easier to understand. But i'll defer to @d-ronnqvist for the call of what is more "Swifty".

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbol.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbol.swift:59]: These aren't platform names here; they're mixin keys, just like the single `Symbol` type:

https://github.com/apple/swift-docc-symbolkit/blob/8a740b5197ece51e4b597f8a467546da66324581/Sources/SymbolKit/SymbolGraph/Symbol/Symbol.swift#L107-L112

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:210]: We should not keep the extra overload groups in the unified symbol graph, because they do not reflect the overloaded symbols of the unified graph. For example, if we kept these extra overload groups in the graph, then Swift-DocC would create superfluous overload pages, with duplicate information, that is already reflected in the "canonical" overload group.

This is the primary reason i did this work in SymbolKit rather than in Swift-DocC: If i had to filter these pages after the fact, i would have to either perform this filtering before all the information is available in DocC's data structures, or else violate invariants about data structures that should not be modified to add or remove new nodes after they've been registered. Overload group symbols are by definition "not real symbols": They do not reflect actual API from the source code. They instead reflect a grouping of other "real" symbols that can be used as a convenience for display. If more than one overload group references the same set of symbols, then that creates more of the clutter in an API view that we are trying to clean up with this project in the first place.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:227]: See my previous comment:

> Overload group symbols are by definition "not real symbols": They do not reflect actual API from the source code. They instead reflect a grouping of other "real" symbols that can be used as a convenience for display.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:231]: Both of these `String` types are referencing symbol unique identifiers. I feel like adding type aliases here would add more clutter than it's worth, especially since this is only a variable binding within a function and these aliases would only be written once or twice.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:331]: I had to look this up while writing it so i could be sure. 😅 

The iterator used in a `for ... in` loop is a _copy_ of the collection being iterated over. (More precisely, [Dictionary's iterator is a value type](https://developer.apple.com/documentation/swift/dictionary/iterator) which contains a reference to the original dictionary, which, like the other collections in the Swift standard library, exercises copy-on-write semantics so that the iterator's view of the dictionary doesn't change while the caller's view does.)

By modifying the collection while iterating over it, you don't invalidate the iterator itself. However, you can still create confusing situations, which is why i perform a `guard` at the beginning of the outer loop so that i don't process overload groups that have been removed.

I can see this being cleaned up by creating a new set of `processedOverloadGroups` and checking membership there instead of in the main dictionary, but this is the version i wound up writing. 🤷‍♀️

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:281]: This way i only had to iterate over the overloads once, but i don't see a single overload group having a ton of symbols, so i can clean this up like you suggest. 👍

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:310]: `relationshipsByLanguage` is public API, so changing its type is a breaking change. 😢 The relationships listing can get pretty large, though, so there might be a better way to perform this lookup so that we don't potentially iterate over thousands of items dozens of times. I'll have to take a look.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:238]: The `siblingOverloadGroups.contains` assertion doesn't have anything to do with the fact that the identifiers should reflect being an overload group. It's making sure that the unified graph view of the overloads (and its sorting method) reflected at least one of the constituent symbol graphs' view (and sorting) of the overloaded symbols. If the unified graph computes a different overload group than any of the individual graphs, that reflects a bug in the code. (One that i ran into several times while testing... 🥴)

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:339]: We need to clean up the relationships listings somehow, and like i said, the type signature here is public API. There can probably be improvements made to iterate this collection fewer times (performing the "make sure the computed overload group references all the symbols" check at the same time as the "remove the relationships for the extra overload groups" pass, combining the "extra overload groups" passes together, etc) but iterating the collection is unavoidable.

File: Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift
Comment (patshaughnessy) [reply Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift:165]: Ok makes sense. The force-unwrap itself wouldn't worry me if the allSatisfy check was in the same function. But now it's sort of an undocumented requirement that callers might not know about.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (patshaughnessy) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:331]: Thanks a lot for the explanation! Super helpful. No, this code is fine as along as we know iterators work this way.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (patshaughnessy) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:238]: Got it thank you.

**Comment by QuietMisdreavus** [comment]
@swift-ci Please test

File: Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift
Comment (patshaughnessy) [inline Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift:56]: Nice! Looks great now.

File: Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift
Comment (patshaughnessy) [inline Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift:115]: Thanks for typing this message out - will save someone's time in the future I'm sure!

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (patshaughnessy) [inline Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:314]: Very clever! 👍

**Review by patshaughnessy — APPROVED** [review]
Nice work - thanks for the updates today. I think the sorting code is a lot cleaner 👏

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (mayaepps) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:261]: Ohh never mind, I think this is addressing situation 2 in your comment.  Seems like `overloadedSymbolIdentifiers` would include overloadA, overloadB, overloadC, and overloadD. Thanks again for that comment 😄

**Review by mayaepps — APPROVED** [review]
Thank you for explaining the intricacies of this!

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (d-ronnqvist) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:167]: One question: It's very possible that I don't see the bigger picture here but it seems like all of the situations that needs post processing happen because `SymbolGraph/createOverloadGroupSymbols()` runs on each individual symbol graph and doesn't have the full unified context. 

Assuming, ` combineOverloadGroups()` always runs, would it make sense to defer some of the work from `createOverloadGroupSymbols()` to only do it once? AFAICT it looks like `OverloadData` is recomputed for all overloaded symbols so I _think_ we could skip doing it in `createOverloadGroupSymbols()`.

Also, since you call out that iterating over `relationshipsByLanguage` could have a performance impact, would it make sense to store all the `.overloadOf` relationships to the side—that is, not in `SymbolGraph/relationships`—so that `combineOverloadGroups()` could operate on a much smaller list?

That would mean that both places (below) that iterate over all relationships but only care about `.overloadOf` relationships can work on that smaller list:
```swift
for relationship in self.relationshipsByLanguage.flatMap(\.value)
    where relationship.kind == SymbolGraph.Relationship.Kind.overloadOf
{
    ...
}
```

```swift
relationshipsByLanguage[selector] = relationships.filter({ relationship in
    ...
})
```

Then at the end we'd append the filtered down `.overloadOf` relationships to all the relationships and remove the side storage.

Does that make sense?

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:257]: Oh right! I guess i didn't think about stuff like `popFirst()` on Sets since it's hidden away in the docs. 😅 I'll add this suggestion, thanks!

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:167]: I could see creating a `createOverloadGroupSymbols` that runs on unified graphs in addition to the one that runs on individual graphs, but i feel like that creates a lot of duplicate work. We could also do the side-storage suggestion for within the graph collector itself, so that it can collect the `overloadOf` relationships ahead of time since it's going to be iterating over all the symbol graphs' relationships anyway. That way the `combineOverloadGroups` work could just take those pre-collected relationships and not need to do the advance scanning like you said. I'll take a look at working that in.

Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:167]: On the other hand, i think it's worth investigating that in a future PR, so that this one doesn't go back for more rounds of review now that it's gotten a few approvals.

File: Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift:172]: I like that as an option, since it makes it more obvious what's going on here.

File: Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift
Comment (QuietMisdreavus) [reply Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift:183]: Yes and no. Here, it means to add it to the `unifiedMixins`, which is done here by recreating the OverloadData and adding it that way. The unified mixins are meant to reflect combined mixins that aren't specific to a given language/platform, which is accomplished here by the recalculation and combining of overload groups.

**Comment by QuietMisdreavus** [comment]
@swift-ci Please test

**Pull request merged**
